tools/migration.py

```diff
@@ -0,0 +1,248 @@
+            return False
+
+        # some findings don't have a start line
+        has_line1 = "rawStartLine" in finding1["location"]
+        has_line2 = "rawStartLine" in finding2["location"]
+        if has_line1 != has_line2:
+            return False
+        if not has_line1:
+            return True
+        return finding1["location"]["rawStartLine"] == finding2["location"]["rawStartLine"]
+
+    @abstractmethod
+    def migrate(self):
+        pass
+
+
+class BlacklistMigrator(MigratorBase):
+    """ Class for migrating blacklists between two instances.
+    If some blacklisted finding cannot be found in the new instance, they will not
+    be migrated.
+    """
+
+    def migrate(self):
+        """ Migrates the blacklists. """
+        migrate_blacklist = self.get_blacklist_infos()
+        for blacklisted in migrate_blacklist:
+            new_id = self.get_matching_finding_id(blacklisted["findingId"])
+            if not new_id:
+                print("could not find finding %s in new Teamscale" % blacklisted["findingId"])
+            else:
+                print("mapped old finding %s to new finding %s" % (blacklisted["findingId"], new_id))
+                self.blacklist(blacklisted, new_id)
+
+        if len(migrate_blacklist) == 0:
+            print("No new blacklisted findings to migrate.")
+        else:
+            print("migrated %d/%d blacklisted findings" % (self.migrated, len(migrate_blacklist)))
+
+    def get_blacklist_infos(self):
+        """ Returns all blacklist info objects from the old instance. """
+        # Remove findings which have already been migrated and have the same id
+        blacklisted_ids = set(self.get_from_old("finding-blacklist")) - set(self.get_from_new("finding-blacklist"))
+
+        infos = []
+        for finding_id in blacklisted_ids:
+            info = self.get_from_old("finding-blacklist", path_param=finding_id)
+            if not info:
+                print("Blacklisted finding %s no longer exists at HEAD, not migrating" % finding_id)
+            else:
+                infos.append(info)
+        return infos
+
+    def blacklist(self, blacklist_info, new_id):
+        """ Blacklists a finding with the given id on the new instance. """
+        self.migrated += 1
+        blacklist_info["findingId"] = new_id
+        self.put("finding-blacklist", blacklist_info, path_param=new_id)
+
+
+class TaskMigrator(MigratorBase):
+    """ Class for migrating tasks between two instances.
+    Tasks will only be migrated if all connected findings are on the new instance as well.
+    """
+
+    def migrate(self):
+        """ Migrates the tasks. """
+        parameters = {"with-count": True, "max": 0}
+        new_tasks_count = self.get_from_new("tasks", parameters=parameters)["totalTaskCount"] + 1
+
+        old_tasks = self.get_from_old("tasks", parameters={"details": True})
+        total = len(old_tasks)
+        for old_task in old_tasks:
```

Comment:
this loop is a bit convoluted. I'd prefer it if you separated it into two parts: first is filtering the old tasks based on whether they already exist, second is migrating. this way, you can also calculate the total with a simple len() call after filtering, which is much easier to understand